Fix the section on 64bit systems (#356)
w3c/distributed-tracing: spec/60-trace-id-format.md

## Considerations for trace-id field generation

This section suggests some best practices to consider when platform or tracing
vendor implement `trace-id` generation and propagation algorithms. These
practices will ensure better interoperability of different systems.

### Uniqueness of `trace-id`

The value of `trace-id` SHOULD be globally unique. This field is typically used
for unique identification of a <a>distributed trace</a>. It is common for
<a>distributed traces</a> to span various components, including, for example,
cloud services. Cloud services tend to serve variety of clients and have a very
high throughput of requests. So global uniqueness of `trace-id` is important,
even when local uniqueness might seem like a good solution.

### Randomness of `trace-id`

> Randomly generated value of `trace-id` SDHOULD be preferred over other

Reviewer: ```suggestion
Randomly generated value of `trace-id` SHOULD be preferred over other
```

algorithms of generating a globally unique identifiers. Randomness of `trace-id`

> addresses some [secutiry](#security-considerations) and [privacy

Reviewer: ```suggestion
addresses some [security](#security-considerations) and [privacy
```

concerns](#privacy-considerations) of exposing unwanted information. Randomness
also allows tracing vendors to base sampling decisions on `trace-id` field value
and avoid propagating an additional sampling context.

As shown in the next section, it is important for better inter-operability with
some existing systems, for `trace-id` to carry "uniqueness" and "randomness" in

> the right part of the `trace-id`.

Reviewer: swap parts of the sentence

As shown in the next section, it is important for `trace-id` to carry "uniqueness" and "randomness" in the right part of the `trace-id`, for better inter-operability with some existing systems.

### Length of `trace-id` for the platforms operating with shorter identifiers

> There are system operates with a `trace-id` that is shorter than 16 bytes,

Reviewer: ```suggestion
There are system that operate with a `trace-id` that is shorter than 16 bytes,
```

> which still willing to adopt this specification.

Reviewer: ```suggestion
which are still willing to adopt this specification.
```

> If such system capable of propagating an entire `trace-id`, even while operating

Reviewer: ```suggestion
If such a system is capable of propagating an entire `trace-id`, even while operating
```

internally on shorter identifier, system is encouraged to utilize the
`tracestate` header to propagate shorter identifier. However, if system wants to

> use a subset of trace-id bytes as an identifier - rightmost part of a `trace-id`

Reviewer: ```suggestion
use a subset of trace-id bytes as an identifier - the rightmost part of a `trace-id`
```

SHOULD be used.

There are tracing systems that are not capable to propagate the entire 16 bytes
of a `trace-id`. For better interoperability of these systems with the fully
compliant systems, the following practices are recommended:

> 1. Whenever a new `trace-id` needs to be generated - it SHOULD be left padded

Reviewer: I think this whole point is irrelevant. How the system operates internally, including ID generation, is not important, because this specification would only matter on egress from such system.

   with zeroes. For example, for the identifier `53ce929d0e0e4736`, `trace-id`
   must be set to `000000000000000053ce929d0e0e4736`.

> 2. On incoming `traceparent`, the rightmost part of `trace-id` SHOULD be used as

Reviewer: I would split this into two bullet items:
  * on ingress, the incoming trace-id should be truncated by dropping the left-most bytes
  * on egress, the outgoing trace-id should be formed by padding the internal 64bit id with 0 bytes on the left

Reviewer 2: I agree about reframing this section, but I don't see it so much as ingress vs egress, I see it as crossing W3C compliance boundaries.  

When a trace moves from a 64 bit system to a W3C compliant 128 bit system, the 128 bit system should left-pad with zeros.  

When a trace moves from a W3C compliant 128 bit system to a 64 bit system, we expect the 64 bit system to either propagate the full 128 bits (if possible) or propagate only the 64 least-significant bits of the trace id.

Reviewer 3: Just to add a further clarification: the conversions must happen in the new, complaint systems, not in the legacy systems. The legacy systems should be assumed to be unchangeable. So the egress/ingress directions are slightly different.

I also feel that it would be clearer to have two completely separate sections: one for describing new systems which use a non-complaint internal identifier, and one for new systems interoperating with legacy systems.

This would be my suggested change:

```markdown
### Handling `trace-id` for complaint platforms with shorter internal identifiers

There are tracing systems which use a `trace-id` that is shorter than 16 bytes, 
which are still willing to adopt this specification.

If such a system is capable of propagating a fully compliant `trace-id`, even 
while still requiring a shorter, non-complaint identifier for internal purposes, 
the system is encouraged to utilize the `tracestate` header to propagate the 
additional internal identifier. However, if a system would instead prefer to use 
the internal identifier as the basis for a fully complaint `trace-id`, it SHOULD 
be incorporated at the as rightmost part of a `trace-id`.

### Interoperating with legacy formats which use shorter identifiers

There are tracing systems which are not capable of propagating the entire 16 
bytes of a `trace-id`. When a fully compliant system wished to interoperarate 
with these legacy systems, the following practices are recommended:

1. Whenever a shorter `trace-id` is recieved from a legacy system, a compliant 
   `trace-id` SHOULD be generated by left padding the original identifier with 
   zeroes. For example, when receiving the identifier `53ce929d0e0e4736`, the 
   new `trace-id` should be set to `000000000000000053ce929d0e0e4736`.
2. When converting `traceparent` to a legacy format, the rightmost part of 
   `trace-id` SHOULD be used as the identifier when truncating. For instance, 
   if the the value of `trace-id` was `234a5bcd543ef3fa53ce929d0e0e4736` on an 
   incoming request, all outgoing requests to legacy systems would be sent a 
   `trace-id` with the value of `53ce929d0e0e4736`.
```

Author: I think I resolved it https://github.com/w3c/trace-context/pull/356/commits/75f04af39c509ba061d9c64e2979234dfe2274a6

Modified text suggested by @tedsuo a bit. And added a note about `tracestate` propagation

inter-operability with some existing systems.

### Handling `trace-id` for complaint platforms with shorter internal identifiers

There are tracing systems which use a `trace-id` that is shorter than 16 bytes,
which are still willing to adopt this specification.

If such a system is capable of propagating a fully compliant `trace-id`, even
while still requiring a shorter, non-complaint identifier for internal purposes,
the system is encouraged to utilize the `tracestate` header to propagate the
additional internal identifier. However, if a system would instead prefer to use
the internal identifier as the basis for a fully complaint `trace-id`, it SHOULD
be incorporated at the as rightmost part of a `trace-id`. For example, tracing
system may receive `234a5bcd543ef3fa53ce929d0e0e4736` as a `trace-id`, hovewer
internally it will use `53ce929d0e0e4736` as an identifier.

### Interoperating with existing systems which use shorter identifiers

There are tracing systems which are not capable of propagating the entire 16
bytes of a `trace-id`. For better interoperability between a fully compliant
systems with these existing systems, the following practices are recommended:

> 1. Whenever a shorter `trace-id` needs to be send, a compliant `trace-id` SHOULD

Reviewer: sorry, I disagree with @tedsuo's comments on ingress/egress. The responsibility for conversion should be on the non-compliant system, not on the compliant sender/receiver. If that system is a blackbox that cannot be changed, it can be deployed with proxies / sidecar that will perform rewriting of the headers to the format it understands. But if my service is fully compliant and needs to talk to such legacy system, I should not be required to couple my service with understanding of what the legacy system understands or not.

Reviewer: This is why I was suggesting to phrase these in terms of ingress/egress to/from non-compliant region of the architecture.

Author: the word compliant here refers to compliant trace-id. I can replace it with 16 bytes `trace-id`

Author: I think @tedsuo wrote the paragraph with the idea on how to convert from traceparent to some other format with shorter ID. This is why it's harder to read with the different idea in mind. I added paragraph to talk about conversions

Author: added clarifications and rephrased a bit

   be generated by left padding the original identifier with zeroes. For
   example, the identifier `53ce929d0e0e4736`, SHOULD be converted to
   `trace-id` value `000000000000000053ce929d0e0e4736`.
2. When converting `traceparent` to a shorter identifier, the rightmost part of
   `trace-id` SHOULD be used as the identifier. For instance, if the the value
   of `trace-id` was `234a5bcd543ef3fa53ce929d0e0e4736` on an incoming request,
   tracing system SHOULD use identifier with the value of `53ce929d0e0e4736`.

Similar transformations are expected when tracing system converts other
distributed trace context propagation formats to W3C Trace Context. Shorter
identifiers SHOULD be left padded with zeros when converted to 16 bytes
`trace-id` and rightmost part of `trace-id` SHOULD be used as a shorter
identifier.

Note, many existing systems that are not capable of propagating the whole
`trace-id` will not propagate `tracestate` header either. However, such system

> can still use `tracestate` header to propagate additional that is known by this

Reviewer: > propagate additional

missing word?

inter-operability with some existing systems.

### Handling `trace-id` for complaint platforms with shorter internal identifiers

There are tracing systems which use a `trace-id` that is shorter than 16 bytes,
which are still willing to adopt this specification.

If such a system is capable of propagating a fully compliant `trace-id`, even
while still requiring a shorter, non-complaint identifier for internal purposes,
the system is encouraged to utilize the `tracestate` header to propagate the
additional internal identifier. However, if a system would instead prefer to use
the internal identifier as the basis for a fully complaint `trace-id`, it SHOULD
be incorporated at the as rightmost part of a `trace-id`. For example, tracing
system may receive `234a5bcd543ef3fa53ce929d0e0e4736` as a `trace-id`, hovewer
internally it will use `53ce929d0e0e4736` as an identifier.

### Interoperating with existing systems which use shorter identifiers

There are tracing systems which are not capable of propagating the entire 16
bytes of a `trace-id`. For better interoperability between a fully compliant
systems with these existing systems, the following practices are recommended:

> 1. Whenever a shorter identifier needs to be send, a 16 bytes `trace-id` SHOULD

Reviewer: When a system creates an outbound message and needs to generate a fully compliant 16 bytes `trace-id` from a shorter identifier, it SHOULD left pad the original identifier with zeroes.

> 1. Whenever a shorter identifier needs to be send, a 16 bytes `trace-id` SHOULD

Reviewer 2: I think this description is backwards.  Bullet 1 here is about _upgrading_ a trace id from 64 bits to 128 bits.  

```suggestion
1. To convert an 8 byte `trace-id` to 16 bytes, the `trace-id` SHOULD
```

Author: this is too specific. I took @yurishkuro suggestion that basically conveys the same message. Please review. Thanks!

   be generated by left padding the original identifier with zeroes. For
   example, the identifier `53ce929d0e0e4736`, SHOULD be converted to
   `trace-id` value `000000000000000053ce929d0e0e4736`.

> 2. When converting 16 bytes `trace-id` to a shorter identifier, the rightmost

Reviewer: When a system receives an inbound message and needs to convert the 16 bytes `trace-id` to a shorter identifier, ...

## Considerations for trace-id field generation

This section suggests some best practices to consider when platform or tracing
vendor implement `trace-id` generation and propagation algorithms. These
practices will ensure better interoperability of different systems.

### Uniqueness of `trace-id`

The value of `trace-id` SHOULD be globally unique. This field is typically used

> for unique identification of a <a>distributed trace</a>. It is common for

Reviewer 4: is the `<a>` tag by mistake ```<a>distributed trace</a>```?

Author: no, this is ReSpec thing to point to glossary.
